### PR TITLE
cephfs-shell: return non-zero value on error

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Fixes: https://tracker.ceph.com/issues/42100
Fixes: https://tracker.ceph.com/issues/42101

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
Depends on https://github.com/ceph/ceph/pull/30365.

**Comment by rishabh-d-dave** [comment]
Rebased the PR.

File: src/pybind/cephfs/cephfs.pyx [hunk 1]
```diff
@@ -156,18 +156,15 @@ cdef extern from "cephfs/libcephfs.h" nogil:
 
 
 class Error(Exception):
-    pass
-
-
-class OSError(Error):
```
Comment (batrick) [inline src/pybind/cephfs/cephfs.pyx]: what's the rationale for this change? commit message didn't explain

Comment (batrick) [reply src/pybind/cephfs/cephfs.pyx]: I see, you want this for commit "cephfs-shell: catch libcephfs.Error only in onecmd()". The better thing to do IMO is catch `libcephfs.OSError` instead.

Comment (rishabh-d-dave) [reply src/pybind/cephfs/cephfs.pyx]: > what's the rationale for this change? commit message didn't explain

Wanted `perror(e.strerror)` work for both exceptions. IMO `class Error` accepting different arguments did not make sense. I confirmed about this from Varsha since she changed/defined what args `class Error` takes.

Comment (rishabh-d-dave) [reply src/pybind/cephfs/cephfs.pyx]: > I see, you want this for commit "cephfs-shell: catch libcephfs.Error only in onecmd()". The better thing to do IMO is catch `libcephfs.OSError` instead.

Didn't want to miss out on [`LibCephFSStateError`](https://github.com/ceph/ceph/blob/master/src/pybind/cephfs/cephfs.pyx#L204)

Comment (batrick) [reply src/pybind/cephfs/cephfs.pyx]: > Didn't want to miss catching [`LibCephFSStateError`](https://github.com/ceph/ceph/blob/master/src/pybind/cephfs/cephfs.pyx#L204).

Okay. But I still don't think making the `libcephfs.Error` class always have a strerror/errno field is correct. I think the better approach is to use the __str__ method to generate a message. I'll point elsewhere in this review.

File: src/tools/cephfs/cephfs-shell [hunk 2]
```diff
@@ -405,12 +408,21 @@ class CephFSShell(Cmd):
...
         except ConnectionError as e:
-            self.perror('***\n{}'.format(e), end='\n', apply_style=True)
+            perror('***\n{}'.format(e))
+            self.exit_code = 3
         except KeyboardInterrupt:
-            self.perror('Command aborted', end='\n', apply_style=True)
+            perror('Command aborted')
+            self.exit_code = 130
+        except libcephfs.Error as e:
+            perror(e.strerror)
```
Comment (batrick) [inline src/tools/cephfs/cephfs-shell]: `perror(e)`

cmd2 will show that it's an exception and print the exception using `str(e)`:

````
CephFS:~/>>> mkdir foo/bar
EXCEPTION of type 'ObjectNotFound' occurred with message: 'error in mkdir foo/bar: No such file or directory [Errno 2]'
To enable full traceback, run the following command:  'set debug true'
````

Which looks fine to me. This also allows us to have future exceptions that do not have an errno but **do** have a `__str__` method.

File: src/tools/cephfs/cephfs-shell [hunk 3]
```diff
@@ -130,9 +131,9 @@ def ls(path, opts=''):
                 elif almost_all and dent.d_name in (b'.', b'..'):
                     continue
                 yield dent
-    except libcephfs.ObjectNotFound:
-        perror('{}: no such directory exists'.format(path), end='\n',
-               apply_style=True)
+    except libcephfs.ObjectNotFound as e:
+        perror(e.strerror)
```
Comment (batrick) [inline src/tools/cephfs/cephfs-shell]: Here and everywhere else we should just `perror(e)` IMO.

**Comment by rishabh-d-dave** [comment]
I need to rebase this PR since #30365 has been updated. I'll do that after #30365 merges.

**Comment by batrick** [comment]
Remind me to look at this once #30365 is merged and this is rebased.

**Comment by rishabh-d-dave** [comment]
@batrick
> Remind me to look at this once #30365 is merged and this is rebased.

Sure. But in the mean time can you please review commits on this PR that are absent on #30365, so that this PR moves a step towards merge?

The commits are: 16b1e0bd76e9492129b24992843bf401cf4d8ef7, 62141a5eefb6375badc907093440013121a6a0f2, 96a970b0989edd54b6727ca81cfaa006965aa1c2 and 683c8a39387106a0ffb840f5389f4cc133ee1fa3.

**Comment by rishabh-d-dave** [comment]
@batrick Please review this PR. Teuth tests passed - http://pulpito.front.sepia.ceph.com/rishabh-2019-11-04_14:22:46-fs-wip-rishabh-fs-shell-retval-distro-basic-smithi/.
